dplyr 1.1.0 pre-release post (#609)
tidyverse/tidyverse.org: content/blog/dplyr-1-1-0-is-coming-soon/index.Rmd

-   New join features, such as non-equi joins

-   `arrange()` improvements with character vectors

-   `morph()`, a generalization of `summarise()`

-   Rewrites of "vector" functions using vctrs, including exciting updates to `case_when()`

This pre-release blog post will discuss these new features in more detail.
We feel pretty confident about them and can't wait for you to try them out, but we welcome any feedback on the [dplyr issues page](https://github.com/tidyverse/dplyr/issues).

You can see a full list of changes in the [release notes](https://dplyr.tidyverse.org/dev/news/index.html).
There are many additional improvements that couldn't fit in a single blog post!

dplyr 1.1.0 is not on CRAN yet, but you can install the development version from GitHub with:

```{r, eval = FALSE}
pak::pak("tidyverse/dplyr")
```

The development version is mostly stable, but is still subject to minor changes before the official release.
We don't encourage relying on it for production usage, but we would love for you to try out these new features.

```{r setup, warning=FALSE, message=FALSE}
library(dplyr)
library(clock)
set.seed(12345)
```

## Temporary grouping with `.by`

> Verbs that work "by group," such as `mutate()`, `summarise()`, `filter()`, and `slice()`, have gained a powerful new argument, `.by`, which allows for inline and temporary grouping.

Reviewer: ```suggestion
Verbs that work "by group," such as `mutate()`, `summarise()`, `filter()`, and `slice()`, have gained a experimental new argument, `.by`, which allows for inline and temporary grouping.
```

Grouping radically affects the computation of the dplyr verb you use it with, and one of the goals of `.by` is to allow you to place that grouping specification alongside the code that actually uses it.
As an added benefit, with `.by` you no longer need to remember to `ungroup()` after `summarise()`, and `summarise()` won't ever message you about how it's handling the groups!

> This exciting feature was inspired by [data.table](https://cran.r-project.org/package=data.table), which bakes this idea directly into `[`.

Reviewer: ```suggestion
This feature was inspired by [data.table](https://cran.r-project.org/package=data.table), which has always used per-operation grouping.
```

We'll explore `.by` with this `expenses` dataset, containing various `cost`s tracked across `id` and `region`.

```{r}
expenses <- tibble(
  id = c(1, 2, 1, 3, 1, 2, 3),
  region = c("A", "A", "A", "B", "B", "A", "A"),
  cost = c(25, 20, 19, 12, 9, 6, 6)
)
expenses
```

If I were to ask you to compute the average `cost` per `region`, you'd probably write something like:

```{r}

> expenses %>%

Reviewer: Base pipe?

  group_by(region) %>%
  summarise(cost = mean(cost))
```

With `.by`, you can now write:

```{r}
expenses %>%
  summarise(cost = mean(cost), .by = region)
```

These two particular results look the same, but the behavior of `.by` diverges from `group_by()` when multiple group columns are involved:

-   `summarise(.by = )` always returns an ungrouped data frame

-   `group_by() %>% summarise()` peels off 1 layer of grouping by default, and messages you that it is doing so unless you set `.groups`

```{r}

> # Typically you'd follow this up with `ungroup()` or you'd set `.groups = "drop"`

Reviewer: Probably don't need quite so much code here

Author: I think the example with multiple groups is important, so I've kept that, but I removed the bullets and generally simplified the language because I think the output tells most of the story

expenses %>%
  group_by(id, region) %>%
  summarise(cost = mean(cost))
```

```{r}
expenses %>%
  summarise(cost = mean(cost), .by = c(id, region))
```

You might also recognize that these results aren't returned in exactly the same order.
`group_by()` always sorts the grouping keys in ascending order, but `.by` retains the original ordering found in the data.
If you need ordered summaries with `.by`, we recommend calling `arrange()` explicitly before or after summarizing.

> While we've focused on using `.by` with `summarise()`, it also works with other verbs, like `mutate()` and `slice_max()`:

Reviewer: ```suggestion
While here we've focused on using `.by` with `summarise()`, it also works with other verbs, like `mutate()` and `slice_max()`:
```

Non-equi joins were requested back in 2016, and were the highest requested dplyr feature at the time they were finally implemented, with over [147 thumbs up](https://github.com/tidyverse/dplyr/issues/2240)!
data.table has had support for non-equi joins for many years, and their implementation greatly inspired the one used in dplyr.

To demonstrate the different types of non-equi joins, imagine that you are in charge of the party planning committee for your office.
Unfortunately, you only get to have one party per quarter, but it is your job to ensure that every employee is assigned to a single party.
Upper management has provided the following 4 party dates:

```{r}
parties <- tibble(
  q = 1:4,
  party = date_parse(c("2022-01-10", "2022-04-04", "2022-07-11", "2022-10-03"))
)

parties
```

With this set of employees:

```{r}
employees <- tibble(
  name = wakefield::name(100),
  birthday = date_parse("2022-01-01") + (sample(365, 100, replace = TRUE) - 1)
)

employees
```

One way to start approaching this problem is to look for the party that happened directly before each birthday.
You can do this with an inequality join:

```{r}

> left_join(employees, parties, join_by(birthday >= party))

Reviewer: Use the pipe here?

```{r, include=FALSE}
options(lifecycle_verbosity = "warning")
```

```{r}
df %>%
  summarise(x = intersect(x, table), .by = g)
```

```{r, include=FALSE}
options(lifecycle_verbosity = NULL)
```

That said, we still believe that this is a powerful tool, so we've moved these features to a new verb, `morph()`.
Think of `morph()` as a generic tool for "doing something to each group," with no restrictions on the number of rows returned per group.

```{r}
df %>%
  morph(x = intersect(x, table), .by = g)
```

One big difference between `summarise()` and `morph()` is that `morph()` always returns an ungrouped data frame, even if the input was a grouped data frame with multiple groups.
This simplifies `morph()` immensely, as it doesn't need to inherit the `.groups` argument of `summarise()`, and never emits any messages.

We expect that you'll continue to use `summarise()` much more often than `morph()`, but if you ever find yourself applying a function to each group that returns an arbitrary number of rows, `morph()` should be your go-to tool!

> ## Vector function rewrites

Reviewer: I'd be tempted to omit this since I'm not sure it needs comment. (You can still use it in the release blog post)

Author: Are you talking about the little paragraph that follows this, or this whole section? Did you not want to talk about `case_when()` in the pre-release post? I intended for this to be an H2 header to set up the three H3 headers about `case_when` `case_match` and `consecutive_id`

Reviewer: I was thinking the whole section.

Author: I removed it
